Public/Src/Engine/Processes/SharedOpaqueJournal.cs

```diff
@@ -0,0 +1,183 @@
+        /// <see cref="Pip.FormattedSemiStableHash"/> is used as journal base name.
+        ///
+        /// <seealso cref="SharedOpaqueJournal(PathTable, IEnumerable{AbsolutePath}, AbsolutePath)"/>
+        /// </summary>
+        public SharedOpaqueJournal(PipExecutionContext context, Process process, AbsolutePath journalDirectory)
+            : this(
+                  context.PathTable, 
+                  process.DirectoryOutputs.Where(d => d.IsSharedOpaque).Select(d => d.Path),
+                  GetJournalFileForProcess(context.PathTable, journalDirectory, process))
+        {
+            Contract.Requires(process != null);
+            Contract.Requires(context != null);
+            Contract.Requires(journalDirectory.IsValid);
+            Contract.Requires(Directory.Exists(journalDirectory.ToString(context.PathTable)));
+        }
+
+        /// <summary>
+        /// Creates a new journal.
+        /// </summary>
+        /// <param name="pathTable">Path table.</param>
+        /// <param name="rootDirectories">Only paths under one of the root directories are recorded in <see cref="RecordFileWrite(AbsolutePath)"/>.</param>
+        /// <param name="journalPath">File to which to write recorded accesses.</param>
+        public SharedOpaqueJournal(PathTable pathTable, IEnumerable<AbsolutePath> rootDirectories, AbsolutePath journalPath)
+        {
+            Contract.Requires(pathTable.IsValid);
+            Contract.Requires(rootDirectories != null);
+
+            m_pathTable = pathTable;
+            m_rootDirectories = rootDirectories.ToArray();
+
+            JournalPath = journalPath.ToString(pathTable);
+            m_bxlWriter = new BuildXLWriter(
+                stream: new FileStream(JournalPath, FileMode.Create, FileAccess.Write, FileShare.Read | FileShare.Delete),
+                debug: false,
+                logStats: false,
+                leaveOpen: false);
+        }
+
+        /// <summary>
+        /// Given a root directory (<paramref name="journalDirectory"/>), returns the full path to the journal corresponding to process <paramref name="process"/>.
+        /// </summary>
+        public static AbsolutePath GetJournalFileForProcess(PathTable pathTable, AbsolutePath journalDirectory, Process process)
+        {
+            Contract.Requires(journalDirectory.IsValid);
+            return journalDirectory.Combine(pathTable, process.FormattedSemiStableHash);
+        }
+
+        /// <summary>
+        /// Finds and returns all journal files that exist in directory denoted by <paramref name="directory"/>
+        /// </summary>
+        public static IEnumerable<string> FindAllJournalFiles(string directory)
+        {
+            return Directory.Exists(directory)
+                ? Directory.EnumerateFiles(directory, "*", SearchOption.TopDirectoryOnly)
+                : CollectionUtilities.EmptyArray<string>();
+        }

```

Comment:
This looks slightly weird here, since there is in fact no logic to distinguish a journal file from one that is not. If I'm getting this right, the assumption is that all journal files are placed flat under a well-known directory.
So it doesn't look like the responsibility of this class to know that. I'd move this logic to whoever decides to put all journal files under the same directory (I guess it is EngineSchedule)

Reply:
indeed

Reply:
In the end I left it here because it should be consistent with `GetJournalFileForProcess`.

Here's what I did to make it better:
  - `GetJournalFileForProcess` returns a path like "Pip{SemiStableHash}.journal"
  - this method searches for all "Pip*.journal" files
  - added a CODESYNC comment